Setup pac config syncer

sm43 opened the pull request:
this setups a configmap syncer which uses informer and loooks for configmap changes and update the run.Info.

# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

sm43 committed:
Setup pac config syncer

this setups a configmap syncer which uses informer and loooks for
configmap changes and update the run.Info.

chmouel commented:
/test go-testing

sm43 commented:
> /test go-testing

![Screenshot 2024-04-15 at 15 46 29](https://github.com/openshift-pipelines/pipelines-as-code/assets/55777192/31626aab-5d90-4198-ae52-8e2f0d7bc57c)

is this a bug? no change in the icon ? and Failing status

chmouel commented:
was the issue, when it failed first it didn't change the icon and kept as failed until succeeded right?

 we merged some change related to that on friday cc @savitaashture

sm43 commented:
> was the issue, when it failed first it didn't change the icon and kept as failed until succeeded right?

yes. when it was restarted using /test. the status icon didn't changed to in-progress

chmouel commented:
/test go-testing

chmouel commented:
it seems to work here 
![image](https://github.com/openshift-pipelines/pipelines-as-code/assets/98980/ea7ec9ff-15b8-4bd7-8aee-b5a365b67545)

maybe we should get it to fail first 🤔

sm43 commented:
/test linters

sm43 commented:
![Screenshot 2024-04-15 at 17 50 23](https://github.com/openshift-pipelines/pipelines-as-code/assets/55777192/094c11bb-8422-4c2e-ab4c-3a2088c240b2)


same behaviour .. CI has started but no change in status icon

sm43 commented:
/cancel go-testing

chmouel commented:
@savitaashture can you have a look ? i think it's related to this commit https://github.com/openshift-pipelines/pipelines-as-code/commit/1fe9f45d7d35eea6915471450fd6e2dd145f583a

chmouel commented:
I have done some stress testing and i didn't see any crash 🍾 

I did try the feature and it has been working fine:

i change the value twice and that worked (or at least from the log)

```console
💡 06:55:36 pac-controller updating value for field MaxKeepRunsUpperLimit: from '0' to '2'
💡 06:55:46 pac-controller updating value for field MaxKeepRunsUpperLimit: from '2' to '10'
```

but if i delete the value and then added back after that doesn't seem to get detected

ie: if i delete   `max-keep-run-upper-limit` and then edit and add a non default    `max-keep-run-upper-limit: "10"`  it will not detect something new (or at least there is no logging)

sm43 committed:
Settings: set as empty/false/0 in case default value is not available

when a config value is removed and no default is available then set it as
empty string, or zero or false depending on field type.

sm43 commented:
> 
> ie: if i delete `max-keep-run-upper-limit` and then edit and add a non default `max-keep-run-upper-limit: "10"` it will not detect something new (or at least there is no logging)

Yeah. it was a bug. updated the code

chmouel commented:
This looks good to me, thanks a bunch cc @enarha for info..

i'll go ahead and merge and squash unless you wan tto do it ?

savitaashture commented:
> yes. when it was restarted using /test. the status icon didn't changed to in-progress

@sm43 can you give me the steps which you tried

I know the issue as that i made it intentionally for some reason
